### ISSUE TITLE
Is it necessary to implement hash/compare trait and for Array?

### ISSUE THREAD

**Issue by Lampese** [moonbitlang/core#938]
- [x] Compare (done in #1003) (but to be discussed)
- [ ] Hash (to be discussed)

**Comment by bobzhang** [comment]
compare is needed, not sure about hash though

**Comment by tonyfettes** [comment]
use case for hash: BPE tokenizer. In BPE tokenizer I need to maintain a `Map[(Array[Int], Array[Int]), Array[Int]]` when merging multi-byte pairs.

**Comment by tonyfettes** [comment]
I have another use case: hash cons when compiling a computation graph.

The const in a computation graph is a `FixedArray[Float]` (or `Array[Float]`) and I need it to be hashable so that I can derive Hash for my graph data type.

**Comment by bobzhang** [comment]
I agree we should add hash for Array, user can always misuse HashMap if they want
